components/textlocal/sources/new-contact/new-contact.js

```diff
@@ -0,0 +1,99 @@
+const common = require("../common/timer-based");
+
+module.exports = {
+  ...common,
+  key: "textlocal-new-contact",
+  name: "New Contact",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+    groupId: {
+      type: "string",
+      label: "Contact Group",
+      description: "The contact group to monitor for new contacts",
+      async options(context) {
+        const { page } = context;
+        if (page !== 0) {
+          return [];
+        }
+
+        const { groups } = await this.textlocal.getContactGroups();
+        const options = groups.map((group) => ({
+          label: group.name,
+          value: group.id,
+        }));
+        return {
+          options,
+        };
+      },
+    },
+  },
+  methods: {
+    ...common.methods,
+    _isContactProcessed(contact) {
+      const { number } = contact;
+      return Boolean(this.db.get(number));
+    },
+    _markContactAsProcessed(contact) {
+      const { number } = contact;
+      this.db.set(number, true);
+    },
+    async takeContactGroupSnapshot() {
+      const contactScan = await this.textlocal.scanContactGroup({
+        groupId: this.groupId,
+      });
+
+      for await (const contact of contactScan) {
+        this._markContactAsProcessed(contact);
+      }
+    },
+    generateMeta({
+      contact,
+      ts,
+    }) {
+      const {
+        number,
+        first_name: firstName,
+        last_name: lastName,
+      } = contact;
+      const maskedName = this.getMaskedName({
+        firstName,
+        lastName,
+      });
+      const maskedNumber = this.getMaskedNumber(number);
+      const summary = `New contact: ${maskedName} (${maskedNumber})`;
+      return {
+        id: number,
+        summary,
+        ts,
+      };
+    },
+    async processEvent(event) {
+      const isInitialized = this.db.get("isInitialized");
+      if (!isInitialized) {
+        await this.takeContactGroupSnapshot();
+        this.db.set("isInitialized", true);
+        return;
```

Comment:
If this is the first time the source runs, and it isn't initialized, is there a reason why you `return` early? That is, on the first run, is there still real data that should be processed and emitted?

Reply:
I wanted to take the same approach we took with Google Sheets to avoid timing out during `activate()`.
I could take the snapshot during the activation hook but that could cause a timeout in case of big contact lists.

So the first time this event is triggered, we'd take the intial snapshot. At that point, we don't know which contacts were created between during the time the event source was activated and the time it was triggered for the first time (since the TextLocal API does not sort the results or provide timestamps for us to tell which contacts are new vs old). The only way to tell whether a contact is new or not is to check if it's currently present in the contact list but is not in the DB.

This means that the first execution of the event source must be "sacrificed" to collect information about the complete contact list.

I will add a `README.md` to this component so that users are aware of this.